### PR TITLE
move configuration into `pyproject.toml`

### PR DESCRIPTION
**Pull request by zacharyburnett**
aligning with PEP621 and consolidation of configurations into `pyproject.toml`

### CODE REVIEW

**Comment by zacharyburnett** [comment]
waiting for https://github.com/astropy/extension-helpers/pull/48 to be merged to use the `pyproject.toml` interface for `extension-helpers`

File: tox.ini [hunk 1]
```diff
@@ -96,9 +98,8 @@ extras =
 
 commands =
     pip freeze
-
     pytest --pyargs photutils {toxinidir}/docs \
-    cov: --cov photutils --cov-config={toxinidir}/setup.cfg --cov-report xml:{toxinidir}/coverage.xml \
+    cov: --cov photutils --cov-config={toxinidir}/pyproject.toml --cov-report coverage.xml --cov-report term-missing \
```
Comment (larrybradley) [inline tox.ini:102]: This is causing the CI test to fail:  https://github.com/astropy/photutils/actions/runs/6161981964/job/16723173426?pr=1488#step:6:97

**Comment by larrybradley** [comment]
RTD is also failing:  https://readthedocs.org/projects/photutils/builds/21900620/

File: pyproject.toml [hunk 2]
```diff
@@ -1,15 +1,193 @@
...
+]
+"photutils.datasets" = [
+    "data/*",
+]
+"photutils.detection.tests" = [
+    "data/*",
+]
+"photutils.isophote.tests" = [
+    "data/*",
+]
```
Comment (larrybradley) [inline pyproject.toml:124]:
```suggestion
]
"photutils.psf.tests" = [
    "data/*",
]
```

This was added after your initial draft PR.

File: tox.ini (hunk 1 above)
Comment (larrybradley) [reply tox.ini]: I think the `{toxinidir}` path might also be needed here.  We'll see what CI says.

Comment (larrybradley) [reply tox.ini]: Indeed, the coverage file is not found:
https://github.com/astropy/photutils/actions/runs/6162499225/job/16724233090?pr=1488#step:7:70

File: tox.ini [hunk 3]
```diff
@@ -96,9 +98,8 @@ extras =
 
 commands =
     pip freeze
-
     pytest --pyargs photutils {toxinidir}/docs \
-    cov: --cov photutils --cov-config={toxinidir}/setup.cfg --cov-report xml:{toxinidir}/coverage.xml \
+    cov: --cov photutils --cov-config={toxinidir}/pyproject.toml --cov-report xml:coverage.xml --cov-report term-missing \
```
Comment (larrybradley) [inline tox.ini:102]:
```suggestion
    cov: --cov photutils --cov-config={toxinidir}/pyproject.toml --cov-report xml:{toxinidir}/coverage.xml --cov-report term-missing \
```

File: docs/conf.py [hunk 4]
```diff
@@ -25,10 +30,9 @@
     sys.exit(1)
 
 # Get configuration information from setup.cfg
-conf = ConfigParser()
-conf.read([os.path.join(os.path.dirname(__file__), '..', 'setup.cfg')])
-setup_cfg = dict(conf.items('metadata'))
-
+with open(Path(__file__).parent.parent / 'pyproject.toml', 'rb') as configuration_file:
+    configuration = tomllib.load(configuration_file)
+metadata = configuration['metadata']
```
Comment (larrybradley) [inline docs/conf.py:35]:
```suggestion
metadata = configuration['project']
```

This should fix the RTD build.

**Comment by larrybradley** [comment]
If you rebase, it'll also fix the `devdeps` failure.

File: pyproject.toml [hunk 5]
```diff
@@ -1,15 +1,196 @@
+[project]
+name = "photutils"
+description = "An Astropy package for source detection and photometry"
+requires-python = ">=3.8"
```
Comment (larrybradley) [inline pyproject.toml:4]:
```suggestion
requires-python = ">=3.9"
```

File: pyproject.toml [hunk 6]
```diff
@@ -1,15 +1,196 @@
...
+    "scikit-image>=0.19.0",
+    "scikit-learn>=1.0",
+    "gwcs>=0.18",
+    "bottleneck",
+    "tqdm",
+    "rasterio",
+    "shapely",
+]
+test = [
+    "pytest-astropy",
```
Comment (larrybradley) [inline pyproject.toml:63]:
```suggestion
    "pytest-astropy>=0.10",
```

File: pyproject.toml [hunk 7]
```diff
@@ -1,15 +1,196 @@
...
 [build-system]
-requires = ['setuptools',
-            'setuptools_scm>=6.2',
-            'cython>=0.29.30',
-            'oldest-supported-numpy',
-            'extension-helpers']
-build-backend = 'setuptools.build_meta'
+requires = [
+    "setuptools>=61.2",
+    "setuptools_scm[toml]>=3.4",
```
Comment (larrybradley) [inline pyproject.toml:81]:
```suggestion
    "setuptools_scm[toml]>=6.2",
```

**Review by larrybradley — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 8]
```diff
@@ -1,15 +1,196 @@
...
+    "wheel",
```
Comment (larrybradley) [inline pyproject.toml:82]:
```suggestion

```

File: pyproject.toml [hunk 9]
```diff
@@ -1,15 +1,196 @@
...
+text_file_format = "rst"
+addopts = "--doctest-rst"
+remote_data_strict = true
+filterwarnings = [
+    "error",
+    "ignore:numpy\\.ufunc size changed:RuntimeWarning",
+    "ignore:numpy\\.ndarray size changed:RuntimeWarning",
+    # numpy 1.24 deprecation from skimage (can remove for skimage >= 0.20)
+    "ignore:`np.bool8` is a deprecated alias for `np.bool_`:DeprecationWarning",
+]
```
Comment (larrybradley) [inline pyproject.toml:150]:
```suggestion
    # TEMP: remove when matplotlib nightly fixes the row_stack deprecations
    "ignore:`row_stack` alias is deprecated:DeprecationWarning",
]
```

File: pyproject.toml [hunk 10]
```diff
@@ -1,15 +1,197 @@
...
+known_first_party = [
+    "photutils",
+    "extension_helpers",
+]
+use_parentheses = true
+
+[tool.build-sphinx]
+edit-on-github = false
+github-project = 'astropy/photutils'
```
Comment (larrybradley) [inline pyproject.toml:194]:
```suggestion
github_project = 'astropy/photutils'
```
This is causing the RTD build to fail.

Comment (zacharyburnett) [reply pyproject.toml:194]: are these still necessary? I've seen usage of `build-sphinx` in `dosc/conf.py`, but as far as I know, the "edit on GitHub" Sphinx extension is deprecated

Comment (larrybradley) [reply pyproject.toml]: These aren't defined in `setup.cfg` -- I thought you added them.  `edit-on-github` can be removed.  This PR used `github_project` in `docs/conf.py` for resolving issue numbers in the changelog:  https://github.com/astropy/photutils/pull/1488/files#diff-85933aa74a2d66c3e4dcdf7a9ad8397f5a7971080d34ef1108296a7c6b69e7e3R150

Is there a better place to define `astropy/photutils` needed there?

File: pyproject.toml [hunk 11]
```diff
@@ -1,15 +1,197 @@
...
+    "sphinx",
+    "sphinx-astropy>=1.6",
+    "scipy>=1.7.0",
+    "matplotlib>=3.5.0",
+    "scikit-image>=0.19.0",
+    "scikit-learn>=1.0",
+    "gwcs>=0.18",
+    "rasterio",
+    "shapely",
+    "tomli; \"python-version\" < \"3.11\"",
```
Comment (larrybradley) [inline pyproject.toml:75]:
```suggestion
    "tomli; python-version < '3.11'",
```
I tracked down the remaining RTD failure to this line (it was also preventing editable installs locally).

**Review by larrybradley — CHANGES_REQUESTED** [review]

**Comment by zacharyburnett** [comment]
thanks for tracking it down! I was very confused where the issue was

**Review by larrybradley — APPROVED** [review]
Everything passes now.  Thanks, @zacharyburnett !  I'll remove `setup.cfg' in a separate PR.

**Pull request merged**